modules/fedimint-wallet-client/src/deposit.rs

```diff
@@ -0,0 +1,373 @@
+    let esplora_client = esplora_client(esplora_server)?;
+    let script = script.clone();
+
+    let transactions = esplora_client
+        .scripthash_txs(&script, None)
+        .await?
+        .into_iter()
+        .map(|tx| tx.to_tx())
+        .collect::<Vec<_>>();
+
+    Ok(transactions)
+}
+
+async fn transition_tx_seen(
+    old_state: DepositStateMachine,
+    btc_transaction: bitcoin::Transaction,
+    out_idx: u32,
+) -> DepositStateMachine {
+    let DepositStateMachine {
+        operation_id,
+        state: old_state,
+    } = old_state;
+
+    match old_state {
+        DepositStates::Created(created_state) => DepositStateMachine {
+            operation_id,
+            state: DepositStates::WaitingForConfirmations(WaitingForConfirmationsDepositState {
+                tweak_key: created_state.tweak_key,
+                btc_transaction,
+                out_idx,
+            }),
+        },
+        state => panic!("Invalid previous state: {state:?}"),
+    }
+}
+
+async fn await_deposit_address_timeout(timeout_at: SystemTime) {
+    if let Ok(time_until_deadline) = timeout_at.duration_since(fedimint_core::time::now()) {
+        tokio::time::sleep(time_until_deadline).await;
+    }
+}
+
+async fn transition_deposit_timeout(old_state: DepositStateMachine) -> DepositStateMachine {
+    assert!(
+        matches!(old_state.state, DepositStates::Created(_)),
+        "Invalid previous state"
+    );
+
+    DepositStateMachine {
+        operation_id: old_state.operation_id,
+        state: DepositStates::TimedOut(TimedOutDepositState {}),
+    }
+}
+
+async fn await_btc_transaction_confirmed(
+    context: WalletClientContext,
+    global_context: DynGlobalClientContext,
+    waiting_state: WaitingForConfirmationsDepositState,
+) -> TxOutProof {
+    loop {
+        // TODO: make everything subscriptions
+        // Wait for confirmation
+        let consensus_height = match global_context
+            .api()
+            .with_module(LEGACY_HARDCODED_INSTANCE_ID_WALLET)
+            .fetch_consensus_block_height()
+            .await
+        {
+            Ok(consensus_height) => consensus_height,
+            Err(e) => {
+                warn!("Failed to fetch consensus height from federation: {e}");
+                sleep(TRANSACTION_STATUS_FETCH_INTERVAL).await;
+                continue;
+            }
+        };
+
+        let confirmation_height = match get_tx_block_height(
+            &context.esplora_server,
+            waiting_state.btc_transaction.txid(),
+        )
+        .await
+        {
+            Ok(confirmation_height) => confirmation_height,
+            Err(e) => {
+                warn!("Failed to fetch confirmation height: {e}");
+                sleep(TRANSACTION_STATUS_FETCH_INTERVAL).await;
+                continue;
+            }
+        };
+
+        if !confirmation_height
+            .map(|confirmation_height| consensus_height > confirmation_height)
```

Comment:
What about finality delay?

Reply:
Ahh I see. This will just loop until finally the `consensus_height > confirmation_height`. First time after the transaction has been confirmed we'll likely `continue` here because we'll still be waiting on finality delay many confirmations. So it's this looping construct that actually does the job.

Reply:
Yes, far from elegant, but should work imo. Ideally I'd like to use electrum long term, which afaik supports subscriptions, but we need a way to convert the txout proofs it gives us to the ones that core and fedimint use.